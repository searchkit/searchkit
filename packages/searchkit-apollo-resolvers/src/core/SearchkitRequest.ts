import dataloader from 'dataloader'
import { Client, ClientOptions } from '@elastic/elasticsearch'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'
import { SearchkitConfig } from '../resolvers'
import QueryManager from './QueryManager'
import { filterTransform } from './FacetsFns'

export interface SearchResponse<T> {
  took: number
  timed_out: boolean
  hits: {
    total: {
      value: number
    }
    max_score: number
    hits: Array<{
      _id: number
      _source: any
      fields?: any
      highlight?: any
      inner_hits?: any
      matched_queries?: string[]
      sort?: string[]
    }>
  }
  aggregations?: any
}

export const mergeESQueries = (queries) =>
  queries.reduce(
    (combinedQuery, partial) => ({
      ...combinedQuery,
      ...partial,
      aggs: {
        ...combinedQuery.aggs,
        ...partial.aggs
      }
    }),
    {}
  )

const keepaliveHttpsAgent = new HttpsAgent()
const keepaliveAgent = new HttpAgent()

export default class SearchkitRequest {
  private dataloader: any
  private client: Client

  constructor(private queryManager: QueryManager, private config: SearchkitConfig) {
    this.client = new Client({
      node: this.config.host,
      agent: () =>
        new URL(this.config.host).protocol === 'http:' ? keepaliveAgent : keepaliveHttpsAgent
    })

    this.dataloader = new dataloader(async (partialQueries) => {
      const baseQuery = {
        size: 0,
        ...(this.queryManager.hasQuery() && this.config.query
          ? { query: this.config.query.getFilter(this.queryManager) }
          : {}),
        post_filter: filterTransform(this.queryManager, this.config.facets)
      }

      const ESQuery = mergeESQueries([baseQuery, ...(partialQueries as any[])])
      const response = await this.executeQuery(ESQuery)

      return partialQueries.map(() => response)
    })
  }

  public async search(partialQuery): Promise<SearchResponse<any>> {
    return this.dataloader.load(partialQuery)
  }

  private async executeQuery(esQuery): Promise<SearchResponse<any>> {
    try {
      const response = await this.client.search<SearchResponse<any>>({
        index: this.config.index,
        body: esQuery
      })

      return response.body
    } catch (e) {
      console.log(e)
    }
  }
}
