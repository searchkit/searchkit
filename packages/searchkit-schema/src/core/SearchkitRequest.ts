import dataloader from 'dataloader'
import { Client } from '@elastic/elasticsearch'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'
import { SearchkitConfig } from '../resolvers'
import ESQueryError from '../utils/ESQueryError'
import { BaseFacet } from '../facets'
import QueryManager from './QueryManager'
import { filterTransform } from './FacetsFns'

export interface SearchResponse<T> {
  took: number
  timed_out: boolean
  hits: {
    total: {
      value: number
    }
    max_score: number
    hits: Array<{
      _id: number
      _source: any
      fields?: any
      highlight?: any
      inner_hits?: any
      matched_queries?: string[]
      sort?: string[]
    }>
  }
  aggregations?: any
}

export type Query = {
  bool?: {
    must?: Array<Record<string, unknown>>
    should?: Array<Record<string, unknown>>
    filter?: Array<Record<string, unknown>>
  }
}

type BaseQuery = {
  size: number
  query?: Query
  post_filter?: Query
}

export const mergeESQueries = (queries) =>
  queries.reduce(
    (combinedQuery, partial) => ({
      ...combinedQuery,
      ...partial,
      aggs: {
        ...combinedQuery.aggs,
        ...partial.aggs
      }
    }),
    {}
  )

const keepaliveHttpsAgent = new HttpsAgent()
const keepaliveAgent = new HttpAgent()

export default class SearchkitRequest {
  private dataloader: any
  private client: Client

  constructor(
    private queryManager: QueryManager,
    private config: SearchkitConfig,
    private baseFilters: Array<Record<string, unknown>>,
    private facets: Array<BaseFacet>
  ) {
    this.client = new Client({
      node: this.config.host,
      agent: () =>
        new URL(this.config.host).protocol === 'http:' ? keepaliveAgent : keepaliveHttpsAgent
    })

    this.dataloader = new dataloader(async (partialQueries) => {
      const query = this.buildQuery(partialQueries)
      const esQuery = this.config.postProcessRequest ? this.config.postProcessRequest(query) : query
      const response = await this.executeQuery(esQuery)
      return partialQueries.map(() => response)
    })
  }

  public async search(partialQuery): Promise<SearchResponse<any>> {
    return this.dataloader.load(partialQuery)
  }

  private buildQuery(partialQueries): BaseQuery {
    const queryFilter =
      this.queryManager.hasQuery() && this.config.query
        ? this.config.query.getFilter(this.queryManager)
        : null

    const baseFiltersQuery = filterTransform(this.queryManager, this.config.filters)
    const combinedBaseFilters = [].concat(this.baseFilters, baseFiltersQuery?.bool?.must || [])
    const query: Query = queryFilter || (combinedBaseFilters.length > 0 ? {} : null)

    if (combinedBaseFilters.length) {
      if (query.bool) {
        Object.assign(query.bool, {
          filter: query.bool.filter?.length
            ? [].concat(query.bool.filter, combinedBaseFilters)
            : combinedBaseFilters
        })
      } else {
        Object.assign(query, { bool: { filter: combinedBaseFilters } })
      }
    }

    const postFilter = filterTransform(this.queryManager, this.facets)

    let highlight
    this.config.hits.highlightedFields?.forEach((field) => {
      if (!highlight) {
        highlight = { fields: {} }
      }
      if (typeof field == 'string') {
        highlight.fields[field] = {}
      } else {
        highlight.fields[field.field] = field.config
      }
    })

    const baseQuery = { size: 0 }

    return mergeESQueries(
      [
        baseQuery,
        query && { query },
        postFilter && { post_filter: postFilter },
        highlight && { highlight },
        ...(partialQueries as any[])
      ].filter(Boolean)
    )
  }

  private async executeQuery(esQuery): Promise<SearchResponse<any>> {
    const debugMode = process.env.DEBUG_MODE === "true"
    try {
      if (debugMode) console.log(JSON.stringify(esQuery, null, 2))
      const response = await this.client.search<SearchResponse<any>>({
        index: this.config.index,
        body: esQuery
      })

      return response.body
    } catch (e) {
      if (debugMode) console.log(JSON.stringify(e, null, 2))

      if (e.meta?.statusCode === 400) {
        throw new ESQueryError(
          `Elasticsearch query failed. Check your custom filters or configuration. Below is the ES Query`,
          esQuery
        )
      } else {
        throw e
      }
    }
  }
}
