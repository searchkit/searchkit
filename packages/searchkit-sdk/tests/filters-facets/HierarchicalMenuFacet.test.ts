import nock from 'nock'
import SearchkitRequest, {
  MultiMatchQuery,
  HierarchicalMenuFacet,
  SearchkitConfig
} from '../../src'
import lvl1Mock from '../__mock-data__/HierarchicalMenuFacet/lvl1.json'
import lvl2Mock from '../__mock-data__/HierarchicalMenuFacet/lvl2.json'
import lvl3Mock from '../__mock-data__/HierarchicalMenuFacet/lvl3.json'

describe('NumericRangeFilter', () => {
  let request
  let lastESRequest

  beforeEach(() => {
    request = null
    lastESRequest = null

    const config: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new HierarchicalMenuFacet({
          fields: ['category_lvl1.keyword', 'category_lvl2.keyword', 'category_lvl3.keyword'],
          identifier: 'categories',
          label: 'Categories'
        })
      ]
    }

    request = SearchkitRequest(config)
  })

  it('No filters, bring back only level 1 facets', async () => {
    nock('http://localhost:9200')
      .post('/movies/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        lastESRequest = body
        return lvl1Mock
      })

    const response = await request.execute({
      facets: true
    })
    expect(response.facets).toMatchSnapshot()
    expect(lastESRequest.aggs.facet_bucket_all.aggs.categories).toMatchInlineSnapshot(`
      Object {
        "aggs": Object {
          "lvl_1": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl1.keyword",
                },
              },
            },
            "filter": Object {
              "match_all": Object {},
            },
          },
        },
        "filter": Object {
          "match_all": Object {},
        },
      }
    `)
    expect(response.summary.appliedFilters).toEqual([])
  })

  it('Filter on level 1 facet', async () => {
    request.setFilters([{ identifier: 'categories', value: 'Coats and Jackets', level: 1 }])

    nock('http://localhost:9200')
      .post('/movies/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        lastESRequest = body
        return lvl2Mock
      })

    const response = await request.execute({
      facets: true
    })
    expect(lastESRequest.aggs.facet_bucket_categories.aggs.categories).toMatchInlineSnapshot(`
      Object {
        "aggs": Object {
          "lvl_1": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl1.keyword",
                },
              },
            },
            "filter": Object {
              "match_all": Object {},
            },
          },
          "lvl_2": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl2.keyword",
                },
              },
            },
            "filter": Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "category_lvl1.keyword": Object {
                        "value": "Coats and Jackets",
                      },
                    },
                  },
                ],
              },
            },
          },
        },
        "filter": Object {
          "match_all": Object {},
        },
      }
    `)
    expect(response.summary.appliedFilters).toEqual([
      {
        display: 'HierarchicalMenuFacet',
        id: 'categories_Coats and Jackets',
        identifier: 'categories',
        label: 'Categories',
        level: 1,
        type: 'HierarchicalValueSelectedFilter',
        value: 'Coats and Jackets'
      }
    ])
    expect(response.facets[0].entries).toHaveLength(10)
    expect(
      response.facets[0].entries.find(({ label }) => label === 'Coats and Jackets').entries
    ).toHaveLength(9)
    expect(response.facets[0].entries.filter(({ entries }) => entries !== null)).toHaveLength(1)
  })

  it('1 lvl1 filter, bring back level 1..3 facets. Lvl2 facets have lvl1 filter applied, lvl3 will have lvl 1..2 filter applied', async () => {
    request.setFilters([
      { identifier: 'categories', value: 'Coats and Jackets', level: 1 },
      { identifier: 'categories', value: 'Leather jackets', level: 2 }
    ])

    nock('http://localhost:9200')
      .post('/movies/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        lastESRequest = body
        return lvl3Mock
      })

    const response = await request.execute({
      facets: true
    })

    expect(lastESRequest.aggs.facet_bucket_categories.aggs.categories).toMatchInlineSnapshot(`
      Object {
        "aggs": Object {
          "lvl_1": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl1.keyword",
                },
              },
            },
            "filter": Object {
              "match_all": Object {},
            },
          },
          "lvl_2": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl2.keyword",
                },
              },
            },
            "filter": Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "category_lvl1.keyword": Object {
                        "value": "Coats and Jackets",
                      },
                    },
                  },
                ],
              },
            },
          },
          "lvl_3": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl3.keyword",
                },
              },
            },
            "filter": Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "category_lvl1.keyword": Object {
                        "value": "Coats and Jackets",
                      },
                    },
                  },
                  Object {
                    "term": Object {
                      "category_lvl2.keyword": Object {
                        "value": "Leather jackets",
                      },
                    },
                  },
                ],
              },
            },
          },
        },
        "filter": Object {
          "match_all": Object {},
        },
      }
    `)
    expect(response.summary.appliedFilters).toMatchInlineSnapshot(`
      Array [
        Object {
          "display": "HierarchicalMenuFacet",
          "id": "categories_Coats and Jackets",
          "identifier": "categories",
          "label": "Categories",
          "level": 1,
          "type": "HierarchicalValueSelectedFilter",
          "value": "Coats and Jackets",
        },
        Object {
          "display": "HierarchicalMenuFacet",
          "id": "categories_Leather jackets",
          "identifier": "categories",
          "label": "Categories",
          "level": 2,
          "type": "HierarchicalValueSelectedFilter",
          "value": "Leather jackets",
        },
      ]
    `)

    const lvl1Facet = response.facets[0].entries.find(({ label }) => label === 'Coats and Jackets')

    expect(lvl1Facet.entries).toHaveLength(9)
    expect(lvl1Facet.entries.find(({ label }) => label === 'Leather jackets').entries).toHaveLength(
      2
    )
    expect(response.facets[0].entries.filter(({ entries }) => entries !== null)).toHaveLength(1)
  })

  it('Ignore level 3 filter because only lvl1 filter has been applied', async () => {
    request.setFilters([
      { identifier: 'categories', value: 'Coats and Jackets', level: 1 },
      { identifier: 'categories', value: 'Leather jackets', level: 3 }
    ])

    nock('http://localhost:9200')
      .post('/movies/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        lastESRequest = body
        return lvl3Mock
      })

    const response = await request.execute({
      facets: true
    })

    expect(lastESRequest.aggs.facet_bucket_categories.aggs.categories).toMatchInlineSnapshot(`
      Object {
        "aggs": Object {
          "lvl_1": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl1.keyword",
                },
              },
            },
            "filter": Object {
              "match_all": Object {},
            },
          },
          "lvl_2": Object {
            "aggs": Object {
              "aggs": Object {
                "terms": Object {
                  "field": "category_lvl2.keyword",
                },
              },
            },
            "filter": Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "category_lvl1.keyword": Object {
                        "value": "Coats and Jackets",
                      },
                    },
                  },
                ],
              },
            },
          },
        },
        "filter": Object {
          "match_all": Object {},
        },
      }
    `)
    expect(response.summary.appliedFilters).toMatchInlineSnapshot(`
      Array [
        Object {
          "display": "HierarchicalMenuFacet",
          "id": "categories_Coats and Jackets",
          "identifier": "categories",
          "label": "Categories",
          "level": 1,
          "type": "HierarchicalValueSelectedFilter",
          "value": "Coats and Jackets",
        },
        Object {
          "display": "HierarchicalMenuFacet",
          "id": "categories_Leather jackets",
          "identifier": "categories",
          "label": "Categories",
          "level": 3,
          "type": "HierarchicalValueSelectedFilter",
          "value": "Leather jackets",
        },
      ]
    `)
  })
})
