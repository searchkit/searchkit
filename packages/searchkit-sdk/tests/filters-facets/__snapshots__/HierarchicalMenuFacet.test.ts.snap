// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericRangeFilter 1 lvl1 filter, bring back level 1..3 facets. Lvl2 facets have lvl1 filter applied, lvl3 will have lvl 1..2 filter applied 1`] = `
Object {
  "_source": Object {
    "includes": Array [
      "actors",
      "writers",
    ],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {},
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "category_lvl1.keyword": "Coats and Jackets",
                    },
                  },
                  Object {
                    "term": Object {
                      "category_lvl2.keyword": "Leather jackets",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
    "facet_bucket_categories": Object {
      "aggs": Object {
        "categories": Object {
          "aggs": Object {
            "lvl_1": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl1.keyword",
                  },
                },
              },
              "filter": Object {
                "match_all": Object {},
              },
            },
            "lvl_2": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl2.keyword",
                  },
                },
              },
              "filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "category_lvl1.keyword": Object {
                          "value": "Coats and Jackets",
                        },
                      },
                    },
                  ],
                },
              },
            },
            "lvl_3": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl3.keyword",
                  },
                },
              },
              "filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "category_lvl1.keyword": Object {
                          "value": "Coats and Jackets",
                        },
                      },
                    },
                    Object {
                      "term": Object {
                        "category_lvl2.keyword": Object {
                          "value": "Leather jackets",
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          "filter": Object {
            "match_all": Object {},
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "category_lvl1.keyword": "Coats and Jackets",
                },
              },
              Object {
                "term": Object {
                  "category_lvl2.keyword": "Leather jackets",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`NumericRangeFilter Filter on level 1 facet 1`] = `
Object {
  "_source": Object {
    "includes": Array [
      "actors",
      "writers",
    ],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {},
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "category_lvl1.keyword": "Coats and Jackets",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
    "facet_bucket_categories": Object {
      "aggs": Object {
        "categories": Object {
          "aggs": Object {
            "lvl_1": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl1.keyword",
                  },
                },
              },
              "filter": Object {
                "match_all": Object {},
              },
            },
            "lvl_2": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl2.keyword",
                  },
                },
              },
              "filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "category_lvl1.keyword": Object {
                          "value": "Coats and Jackets",
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          "filter": Object {
            "match_all": Object {},
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "category_lvl1.keyword": "Coats and Jackets",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`NumericRangeFilter Ignore level 3 filter because only lvl1 filter has been applied 1`] = `
Object {
  "_source": Object {
    "includes": Array [
      "actors",
      "writers",
    ],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {},
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "category_lvl1.keyword": "Coats and Jackets",
                    },
                  },
                  Object {
                    "term": Object {
                      "category_lvl3.keyword": "Leather jackets",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
    "facet_bucket_categories": Object {
      "aggs": Object {
        "categories": Object {
          "aggs": Object {
            "lvl_1": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl1.keyword",
                  },
                },
              },
              "filter": Object {
                "match_all": Object {},
              },
            },
            "lvl_2": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl2.keyword",
                  },
                },
              },
              "filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "category_lvl1.keyword": Object {
                          "value": "Coats and Jackets",
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          "filter": Object {
            "match_all": Object {},
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "category_lvl1.keyword": "Coats and Jackets",
                },
              },
              Object {
                "term": Object {
                  "category_lvl3.keyword": "Leather jackets",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`NumericRangeFilter No filters, bring back only level 1 facets 1`] = `
Object {
  "_source": Object {
    "includes": Array [
      "actors",
      "writers",
    ],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {
        "categories": Object {
          "aggs": Object {
            "lvl_1": Object {
              "aggs": Object {
                "aggs": Object {
                  "terms": Object {
                    "field": "category_lvl1.keyword",
                  },
                },
              },
              "filter": Object {
                "match_all": Object {},
              },
            },
          },
          "filter": Object {
            "match_all": Object {},
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [],
        },
      },
    },
  },
  "from": 0,
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`NumericRangeFilter No filters, bring back only level 1 facets 2`] = `
Array [
  Object {
    "display": "HierarchicalMenuFacet",
    "entries": Array [
      Object {
        "count": 426,
        "entries": null,
        "label": "T-Shirts",
        "level": 1,
      },
      Object {
        "count": 372,
        "entries": null,
        "label": "Casual Shirts",
        "level": 1,
      },
      Object {
        "count": 324,
        "entries": null,
        "label": "Knitwear",
        "level": 1,
      },
      Object {
        "count": 308,
        "entries": null,
        "label": "Sweats",
        "level": 1,
      },
      Object {
        "count": 231,
        "entries": null,
        "label": "Coats and Jackets",
        "level": 1,
      },
      Object {
        "count": 220,
        "entries": null,
        "label": "Polo shirts",
        "level": 1,
      },
      Object {
        "count": 132,
        "entries": null,
        "label": "Shorts",
        "level": 1,
      },
      Object {
        "count": 117,
        "entries": null,
        "label": "Trousers",
        "level": 1,
      },
      Object {
        "count": 113,
        "entries": null,
        "label": "Swimwear",
        "level": 1,
      },
      Object {
        "count": 80,
        "entries": null,
        "label": "Underwear",
        "level": 1,
      },
    ],
    "identifier": "categories",
    "label": "Categories",
    "type": "HierarchicalMenuFacet",
  },
]
`;
